### ISSUE TITLE
Lint for missing hashCode implementation.

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#72]
Moved from [dartbug.com/22998](http://dartbug.com/22998):

**What steps will reproduce the problem?**
```
class foo {
  final int x;
  foo(this.x);
  bool operator== (foo other) {
    return this.x == other.x;
  }
}

void main() {
  var f1 = new foo(1);
  var f2 = new foo(2);
  print(f1 == f2);
}
```
**What is the expected output? What do you see instead?**

The analyzer does not show a warning or a hint. 
If you compile the code to JavaScript you will get a hint from dart2js:
```
The class 'foo' overrides 'operator==', but not 'get hashCode'.
  bool operator== (foo other) {
               ^^
```
The analyzer should show the same hint.

What version of the product are you using?

Dart Editor version 1.9.1.release (STABLE)
Dart SDK version 1.9.1

**Comment by pq** [comment]
To push this ahead, how about this description:

> **DO** implement `hashCode` when overriding `==`.

Next, I think we need a new category.  It's not a style issue per se.  Perhaps a `BUG` bucket catch-all for programming mistakes?

We also need a larger chunk of text to serve as background rationale.  We might crib something from prior art like [here](http://www.artima.com/lejava/articles/equality.html).  Alternatively, we can cook up something ourselves.

@bwilkerson : thoughts?

**Comment by bwilkerson** [comment]
In CodePro, we had a category for "SDK API usage" that contained audits related to the way code made use of API's defined in core. (We also had other categories for usage of other API's, such as third-party frameworks.)

**Comment by pq** [comment]
:+1: I like that.

**Comment by pq** [comment]
Implemented as `hash_and_equals`.
